### PR TITLE
remove module hack for palettes

### LINKED ISSUE #9248: [FEATURE] Make the palettes module IDE-friendly
Discussion started in #9241.

The issue is that an IDE cannot understand what members exist in the module because of the hack, so no autocompletion and inspection is possible.

The idea is to get rid of the class-module hack and instead just use tuples for palettes.

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #9248
- [ ] tests added / passed
- [x] release document entry (if new feature or API change)

cc @p-himik @mattpap 

Needs a migration guide update and I would like to take the opportunity to add more tests.  I think there's typing things that can be refined as well. Just noting in passing some things we give up:

* can't call `palette.reverse()`
* palette groups are necessarily mutable (no built-in `frozendict`)

### CODE REVIEW

**Comment by bryevdv** [comment]
@p-himik can you confirm this behaves to your needs/expectations in an IDE?

**Comment by mattpap** [comment]
I actually was able to coerce vscode to recognize the module hack by adding `palettes.pyi` file with appropriate type declarations. If this can be used in other IDEs, say pycharm, then this could be a viable stop gap solution, that would allow us to keep lists. Later on maybe we could use PEP 562 to resolve the issue properly (available in 3.7+). I will submit a PR tomorrow for testing.

**Comment by p-himik** [comment]
@bryevdv I have added a couple of comments, but apart from them the overall code seems like it would work. I'll be able to give it a go a bit later.

@mattpap But why? What is the advantage of having lists there? Apart from just using the correct data types, this PR also drastically reduces the size and complexity of the whole module.
And PEP 562 shouldn't be able to really help us here since code like `palettes.some_palette.append(...)` would still be executable. I don't think we should care about users overwriting palettes as a whole (i.e. `palettes.some_palette = [...]`). After all, we don't care about any other module. And why should we care about palettes as well - why should this module be special?

File: bokeh/palettes.py [hunk 1]
```diff
@@ -385,2118 +384,1191 @@
...
-        return ['#0C0786', '#45039E', '#7200A8', '#9B179E', '#BC3685', '#D7566C', '#EC7853', '#FA9D3A', '#FCC726', '#EFF821']
-    @property
-    def Plasma11(self) -> Palette:
-        return ['#0C0786', '#40039C', '#6A00A7', '#8F0DA3', '#B02A8F', '#CA4678', '#E06461', '#F1824C', '#FCA635', '#FCCC25', '#EFF821']
-    @property
-    def Plasma256(self) -> Palette:
-        return [
-            '#0C0786', '#100787', '#130689', '#15068A', '#18068B', '#1B068C', '#1D068D', '#1F058E', '#21058F', '#230590', '#250591', '#270592',
-            '#290593', '#2B0594', '#2D0494', '#2F0495', '#310496', '#330497', '#340498', '#360498', '#380499', '#3A049A', '#3B039A', '#3D039B',
-            '#3F039C', '#40039C', '#42039D', '#44039E', '#45039E', '#47029F', '#49029F', '#4A02A0', '#4C02
```
[the remark this answers is not in the document]
Comment (p-himik) [reply bokeh/palettes.py]: I am not sure why, but for me GitHub shows a preview of a wrong piece of code. In the proper diff view, the comment is at the right place.

File: bokeh/palettes.py [hunk 2]
```diff
@@ -385,2118 +384,1191 @@
...
-#-----------------------------------------------------------------------------
-# Dev API
-#-----------------------------------------------------------------------------
-
-#-----------------------------------------------------------------------------
-# Private API
-#-----------------------------------------------------------------------------
-
-Palette = List[str]
+Palette = Tuple[str]
```
Comment (mattpap) [inline bokeh/palettes.py]: This should be a variadic tuple, i.e. `Tuple[str, ...]`.

File: bokeh/palettes.py [hunk 3]
```diff
@@ -109,7 +109,7 @@
 
 .. code-block:: python
 
-    ['#084594', '#2171b5', '#4292c6', '#6baed6', '#9ecae1', '#c6dbef', '#deebf7', '#f7fbff']
+    ['#084594', '#2171b5', '#4292c6', '#6baed6', '#9ecae1', '#c6dbef', '#deebf7', '#f7fbff')
```
[the remark this answers is not in the document]
Comment (bryevdv) [reply bokeh/palettes.py]: vim actions can be very efficient, but yes I missed a few

**Comment by bryevdv** [comment]
> But why? What is the advantage of having lists there?

They allow users to take palettes and easily modify them in place, reverse them, swap out a single special color, etc. 

> I don't think we should care about users overwriting palettes as a whole (i.e. palettes.some_palette = [...]). After all, we don't care about any other module. And why should we care about palettes as well - why should this module be special?

That's not the motivation here. The only consideration is to prevent palettes themselves from being modified, and that's in result of actual problems that were encountered years ago. At the time we solved that issue with copy-on-access and this PR proposes to solve it with immutable values.

File: bokeh/palettes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bryevdv) [reply bokeh/palettes.py]: I wouldn't want to take on an entire dependency for this, or add new code we have to maintain either. For that effort we can just keep the module hack. I'm really only concerned about the palettes themselves being modified, I think it is far less likely a user would think to modify these dicts so I am OK just leaving them as dicts even if it is slightly less than the ideal.

**Comment by p-himik** [comment]
> that's in result of actual problems that were encountered years ago
@bryevdv Could you point me to those issues? Only if they're easy to find and share, of course.
I just feel like I'm missing something important and I'd like to know about that context.

**Comment by bryevdv** [comment]
@p-himik some of the examples modified palettes, e.g by reversing them or otherwise, and since the docs builds example in the context of a single process, things broke (in subtle and surprising and hard to track down ways). Users reported similar issues.

File: bokeh/palettes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (p-himik) [reply bokeh/palettes.py]: Oh, it's within the comment... Yeah, my IDE doesn't help me here either.

**Comment by p-himik** [comment]
@bryevdv I see. Although I must admit it still sounds more like a usage error than a library error to me. But that's not important.

What's important is that I just tested it and it works just fine! Thank you.

**Comment by bryevdv** [comment]
@philippjfr @jlstevens @jsignell @canavandl we have two separate proposals for this, can you all weigh in?

**Comment by mattpap** [comment]
If someone could try out PR #9384, it would be great, because this discussion may be mute if that PR doesn't work reliably outside my own testing environment.

**Comment by p-himik** [comment]
@mattbest I'm answering here just to keep the conversation in one place.
Seems to work in IntelliJ IDEA just fine, and hence should work in PyCharm as well.

However, apart from the aforementioned concerns, the solution in your proposal has another disadvantage - it's impossible to automatically navigate to the actual data from its usage. Instead, IDEA navigates to the relevant statement in the `pyi` file and I end up having to dig up the correct file and line manually.

**Comment by bryevdv** [comment]
@mattpap thoughts? @p-himik has raised a concern above.  I have a slight preference for this PR. I'd like to get an agreement before putting any more work in, though. (I don't think anyone else is going to chime in)

**Comment by jsignell** [comment]
I think this PR seems very reasonable. I don't anticipate any major issues in holoviews and I like that the palettes module will no longer be special cased.

**Comment by bryevdv** [comment]
OK I will finish this up tonight

**Pull request merged**